Mathlib/Data/Nat/Bitwise.lean

```diff
@@ -164,6 +244,125 @@ theorem testBit_two_pow (n m : ℕ) : testBit (2 ^ n) m = (n = m) := by
     simp [h]
 #align nat.test_bit_two_pow Nat.testBit_two_pow
 
+theorem testBit_translate (h: i < w) : Nat.testBit (2^w * b + n) i = Nat.testBit n i := by
+  simp only [testBit, shiftr_eq_div_pow, bodd_eq_iff]
+  rw [Nat.add_div_of_dvd_right (by simp [Dvd.dvd.mul_right, pow_dvd_pow, le_of_lt h]), add_mod]
+  have h1: (2^w / 2^i) % 2 = 0 := by
+    rw [← Nat.dvd_iff_mod_eq_zero]
+    use 2^(w - (i + 1))
+    rw [Nat.pow_div (by linarith) (by decide), mul_comm, ← pow_succ 2 _, succ_eq_add_one]
+    simp [← Nat.sub_add_comm, succ_le_of_lt h]
+  simp [mul_comm, Nat.mul_div_assoc b (pow_dvd_pow 2 (le_of_lt h)), mul_mod, h1]
+
+theorem testBit_translate' (h: n < 2^w) : Nat.testBit (2^w * b.toNat + n) w = b:= by
+  simp only [Nat.testBit, Nat.shiftr_eq_div_pow]
+  rw [Nat.add_div_of_dvd_right (Dvd.intro _ rfl), Nat.div_eq_zero h, add_zero]
+  cases' b <;> simp 
+
+theorem testBit_translate_one (h : i < w) : 
+  Nat.testBit (2^w + n) i = Nat.testBit n i := mul_one (2^w) ▸ (testBit_translate h)
+
+theorem testBit_translate_one' (h : n < 2^w) : Nat.testBit (2^w + n) w = true :=
+  mul_one (2^w) ▸ toNat_true ▸ (@testBit_translate' n w true h)
+
+/-- Generic method to create a natural number by appending bits tail-recursively.
+It takes a boolean function `f` on each bit and `z` the starting point and the number of bits `i`.
+It is almost always specialized with `z = 0` and `i = w`; the length of the binary representation.
+This is an alternative to using `List` altogether. It will be used for bitadd, bitneg, bitmul etc.-/
+def toNat (f : Nat → Bool) (z : Nat) : Nat → Nat
```

Comment:
The full name of this is `Nat.toNat` I think. It should be called something more descriptive like `Nat.ofBits`, or maybe `ofBits` should be for the version where `z = 0`. I also don't understand the semantics of the function from the docstring, is `z` effectively added to the result?